### PR TITLE
Remove agent messages at the beginning of the list

### PR DESCRIPTION
**Pull request by lasryaric**
## Description

Context [here](https://github.com/dust-tt/tasks/issues/699).

This is a PR to ensure that `renderModelForConverastion()` never returns a list of messages that starts with an agent message.
The reason for this change is that Claude models do not support having an agent message before a user message.

We think it's reasonable to apply this change uniformly for all models.

### CODE REVIEW

File: front/lib/api/assistant/generation.ts
Comment (fontanierh) [inline front/lib/api/assistant/generation.ts]: super nit but you could maintain tokensUsed as you shift no ?

**Review by fontanierh — APPROVED** [review]
Looking good !

File: front/lib/api/assistant/generation.ts
Comment (lasryaric) [reply front/lib/api/assistant/generation.ts]: Indeed, if you prefer it that way, I can change it :)

File: front/lib/api/assistant/generation.ts
Comment (spolu) [inline front/lib/api/assistant/generation.ts:249]: I thiiiink this calculation is incorect since `tokensUsed` is initialized with `let tokensUsed = promptCountRes.value + tokensMargin;` but also edited on l.203 without a selected.unshift.

I would rather update tokenUsed to be tokensUsed in `selected` and keep the latest snapshot and pull the tokensUsed of the last element of `selected` here?

**Review by spolu — CHANGES_REQUESTED** [review]
I thiiink the last calculation of tokens used is incorrect?

File: front/lib/api/assistant/generation.ts
Comment (lasryaric) [reply front/lib/api/assistant/generation.ts:249]: Indeed. I will follow @fontanierh 's advice on only decreasing it when I pop, which is more robust. Thanks for catching it.

Comment (lasryaric) [reply front/lib/api/assistant/generation.ts:249]: s/pop/shift/

File: front/lib/api/assistant/generation.ts
Comment (spolu) [inline front/lib/api/assistant/generation.ts:254]: no need to repeat the variable here?

**Review by spolu — APPROVED** [review]

File: front/lib/api/assistant/generation.ts
Comment (philipperolet) [inline front/lib/api/assistant/generation.ts]:
```suggestion
    tokensCount -= messageCountRes[messages.length - selected.length]
    selected.shift();
```
would this suggestion work? it would allow not having to make all the other modifications to messages and adding a tokencount to each message, limiting the PR to 4 lines?

**Review by philipperolet — COMMENTED** [review]
Thanks! I have a suggestion here

File: front/lib/api/assistant/generation.ts
Comment (philipperolet) [reply front/lib/api/assistant/generation.ts]: Really think it's worth doing

Comment (lasryaric) [reply front/lib/api/assistant/generation.ts]: Looks like it would work but it makes it quite hard to follow and more error prone IMHO.
I am in favor of keeping it as is.

Comment (philipperolet) [reply front/lib/api/assistant/generation.ts]: Ok, we can disagree on that. IMHO it's easier to follow and less error prone: 
1. simpler type for the array (messages VS {messages, tokenUsed}), 
2. the change is concentrated in one place VS 5 changes at different places, with a map & a throw

Comment (philipperolet) [reply front/lib/api/assistant/generation.ts]: This being said, given you already have 2 approvals, I will not block the merge, do as you see fit :+1:

Comment (fontanierh) [reply front/lib/api/assistant/generation.ts]: FWIW I agree that @philipperolet's proposal is simpler

**Review by philipperolet — APPROVED** [review]
LGTM! I appreciate the fact that you took the suggestion into account, thank you :pray:

**Pull request merged**
